pulp_rpm/app/tasks/synchronizing.py

```diff
@@ -1151,34 +1166,78 @@ async def parse_packages(self, primary_xml, filelists_xml, other_xml):
-                        raise ValueError(MIRROR_INCOMPATIBLE_REPO_ERR_MSG)
+                    raise Exception(ERR_MSG.format("PKGIDs"))
+                else:
+                    log.warn(WARN_MSG.format("PKGIDs"))
+            # Check for packages with duplicate NEVRAs
+            if not nevra_warning_triggered and pkg.nevra() in nevras:
+                nevra_warning_triggered = True
+                if self.mirror_metadata:
+                    raise Exception(ERR_MSG.format("NEVRAs"))
+                else:
+                    log.warn(WARN_MSG.format("NEVRAs"))
+
+            pkg_nevra = pkg.nevra()
+
+            nevras.add(pkg_nevra)
+            checksums.add(pkg.pkgId)
+
+            # Check that all packages are within the root of the repo (if in mirror_complete mode)
+            if self.mirror_metadata:
+                uses_base_url = pkg.location_base
+                illegal_relative_path = self.is_illegal_relative_path(pkg.location_href)
+
+                if uses_base_url or illegal_relative_path:
+                    raise ValueError(MIRROR_INCOMPATIBLE_REPO_ERR_MSG)
+
+            # Add any srpms to the skip set
+            if skip_srpms and pkg.arch == "src":
+                package_skip_nevras.add(pkg_nevra)
+
+            # Collect the N highest-version packages, kick out the older ones and add those
+            # to the skip list. Don't include any modular packages in the EVR comparisons
+            # since they may be older and we don't want to skip them, or newer and we don't want to
+            # exclude any nonmodular packages on the basis of the modular package existing.
+            if self.repository.retain_package_versions and pkg_nevra not in modular_artifact_nevras:
+                pkg_evr = RpmVersion(pkg.epoch, pkg.version, pkg.release)
+                latest_packages_by_arch_and_name[pkg.arch][pkg.name].append((pkg_evr, pkg_nevra))
+
+        # Ew, callback-based API, gross. The streaming API doesn't support optionally
+        # specifying particular files yet so we have to use the old way.
+        parser.for_each_pkg_primary(verification_and_skip_callback)
+
+        # Go through the package lists, sort them descending by EVR, ignore the first N and then
+        # add the remaining ones to the skip list.
+        for arch, packages in latest_packages_by_arch_and_name.items():
+            for name, versions in packages.items():
+                versions.sort(key=lambda p: p[0], reverse=True)
+                for pkg in versions[self.repository.retain_package_versions :]:
+                    (evr, nevra) = pkg
+                    package_skip_nevras.add(nevra)
+
+        del latest_packages_by_arch_and_name
+
+        log.debug(
+            "Skipping {} packages due to retain_package_versions".format(len(package_skip_nevras))
+        )
 
+        # TODO: should this reflect the actual # of packages or just the # synced?
+        # Do we need a new progress bar?
```

Comment:
Need feedback on this

Reply:
I think just synced otherwise it can bring questions "why the counter is at 95/100 packages"